### DISCUSSION TITLE
Always set/keep focus on certain TextInput widget & capture pressing enter

### DISCUSSION THREAD

**Discussion by hjlklink** [AllenDang/giu#704] [Q&A] [answered]
As a complete GIU-novice I am using GIU to design a simple window for a text adventure game. It consists of a label at the top which will contain some information about the player's location. Below that is a read-only multiline input widget which will contain all previous user input & responses given by the game. This should behave a bit like the output of a "console/terminal" window. Below that is a label stating "What now?" and an input text widget for the user to input the next command. Border and background are set to the same color as the window background, so it resembles the look of a "command prompt".

In the future I might want to add a menu bar at the top of the screen (above the previously mentioned title label) for loading & saving the current game state etc.

I would like to set the default focus on the input widget, so the user doesn't have to click it to set the focus. Plus I would like to capture when the "enter" key was pressed, to invoke the interpreter to handle the given input and set the focus back to the input widget after the command response has returned.

Is there a way to achieve this?

**Comment by gucio321** [comment]
for enter see #158
to set focus, there should be `SetKeyboardFocusHere` function somewhere (imgui or giu)

**Reply by gucio321** [reply, marked as answer]
```go
package main

import (
	"fmt"
	
	"github.com/AllenDang/giu"
)

var (
	consoleOutput, userInput string = "line1\nline 2\nline 3", ""
)

func loop() {
	giu.SingleWindow().Layout(
		giu.Label("data about user location"),
		giu.InputTextMultiline(&consoleOutput).Flags(giu.InputTextFlagsReadOnly),
		giu.Custom(func() {
			giu.SetKeyboardFocusHere()
		}),
		giu.InputText(&userInput),
		giu.Event().OnKeyPressed(giu.KeyEnter, func() {
			fmt.Print("enter clicked!")
		}),
	)
}

func main() {
	wnd := giu.NewMasterWindow("A simple text game [discussion 704]", 640, 480, 0)
	wnd.Run(loop)
}
```
ok, here is some code but it doesn't catch enter for some reason :thinking:

**Reply by hjlklink** [reply]
Thank you so much. This seems to do the trick. I wasn't aware that there was such a thing as "Custom" function. :-)

**Reply by gucio321** [reply]
yes! there is a CustomWidget and you can do anything in there :smile:

**Reply by hjlklink** [reply]
The enter event is not always fired/caught, it seems. I have it working sometimes, but even without changing any code it doesn't always work in the subsequent runs anymore. Weird.

EDIT: It seems more stable if I remove the "custom" widget/function part.
EDIT2: It seems to work correctly if I click in the textbox, even though it already has focus because of the "custom" widget function.
